components/raftstore/src/store/peer_storage.rs

```diff
@@ -1210,7 +1212,18 @@ where
 
         if self.is_initialized() {
             // we can only delete the old data when the peer is initialized.
-            self.clear_meta(kv_wb, raft_wb)?;
+            let first_index = self.first_index();
+            // It's possible that logs between `last_compacted_idx` and `first_index` are
+            // being deleted in raftlog_gc worker. But it's OK as:
+            // 1. If the peer accepts a new snapshot, it must start with an index larger than
+            //    this `first_index`;
+            // 2. If the peer accepts new entries after this snapshot or new snapshot, it must
+            //    start with the new applied index, which is larger than `first_index`.
+            // So new logs won't be deleted by on going raftlog_gc task accidentally.
+            // It's possible that there will be some logs between `last_compacted_idx` and
+            // `first_index` are not deleted. So a cleanup task for the range should be triggered
+            // after applying the snapshot.
```

Comment:
Wasn't `last_compacted_idx` and `first_index` are the same? namely `truncated_state().get_index() + 1`

Reply:
`last_compacted_idx` is 0 when initialized.

Reply:
No, it should be scheduled not earlier than the persistent of the deletions.

Reply:
The logs between `last_compacted_idx` and `first_index` should be stale logs that won't be accessed anymore and safe to delete whenever we can.